File: app/src/main/java/com/billiamram/hexagonalscaffold/AndroidExampleModelRepository.java
package com.billiamram.hexagonalscaffold;

import android.content.SharedPreferences;

import com.billiamram.main.ExampleModel;
import com.billiamram.main.ExampleModelRepository;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.List;

public class AndroidExampleModelRepository implements ExampleModelRepository {

    private static final String MODELS_KEY = "models";
    private final Gson mGson;

    private SharedPreferences mSharedPreferences;
    private Type mModelListType = new TypeToken<List<ExampleModel>>() {
    }.getType();

    public AndroidExampleModelRepository(SharedPreferences sharedPreferences, Gson gson) {
        mSharedPreferences = sharedPreferences;
        mGson = gson;
    }

    @Override
    public List<ExampleModel> findAll() {
        String modelsJson = mSharedPreferences.getString(MODELS_KEY, "[]");
        return mGson.fromJson(modelsJson, mModelListType);
    }

    @Override
    public void save(ExampleModel exampleModel) {
        List<ExampleModel> exampleModels = findAll();
        exampleModels.add(exampleModel);

        String modelsJson = mGson.toJson(exampleModels, mModelListType);

        mSharedPreferences.edit().putString(MODELS_KEY, modelsJson).apply();
    }
}

File: app/src/main/java/com/billiamram/hexagonalscaffold/AndroidExampleModelRepository.kt
package com.billiamram.hexagonalscaffold

import android.content.SharedPreferences
import com.billiamram.main.ExampleModel
import com.billiamram.main.ExampleModelRepository
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

class AndroidExampleModelRepository(private val mSharedPreferences: SharedPreferences, private val mGson: Gson) : ExampleModelRepository {
    private val mModelListType = object : TypeToken<List<ExampleModel>>() {}.type

    override fun findAll(): List<ExampleModel> {
        val modelsJson = mSharedPreferences.getString(MODELS_KEY, "[]")
        return mGson.fromJson<List<ExampleModel>>(modelsJson, mModelListType)
    }

    override fun save(exampleModel: ExampleModel) {
        val modelsJson = mGson.toJson(findAll().plus(exampleModel), mModelListType)

        mSharedPreferences.edit().putString(MODELS_KEY, modelsJson).apply()
    }

    companion object {
        private val MODELS_KEY = "models"
    }
}
